### PR TITLE
feat(es): Render and create span data-stream composable templates

### PR DESCRIPTION
**Pull request by Me-Priyank**
## Which problem is this PR solving?

Part of #4708 (RFC 0004 §3.2, Phase 2 milestone 9).

Data streams are declared by a composable index template, which Jaeger has no way to render or install today. This adds that.

## Description of the changes

`IndicesClient.CreateSpanDataStreamTemplates` installs the objects backing the span data stream:

- `<prefix>jaeger.spans@mappings` and `<prefix>jaeger.spans@settings` component templates
- the user-owned `<prefix>jaeger.spans@custom` component
- the composable index template composing all three with `"data_stream": {}`

Both component bodies are **derived** from the existing span index template rather than restated, so shards, replicas and mappings cannot drift between the rotation and data-stream paths. `renderBackendNeutralBody` is extracted for that: the two callers share one replica check, one template execution and one JSON decode, so both fail identically on the same input. It fills the index prefix, shards and replicas from the index config, and takes only the lifecycle settings that differ between them.

`@timestamp` is mapped `date_nanos` because span start times carry microsecond resolution and a plain `date` stores milliseconds (§3.3).

**`@custom` is created when absent, and the write is conditional.** Referencing it without creating it does not survive the backend matrix: `ignore_missing_component_templates` arrived in Elasticsearch 8.7 and exists in no OpenSearch version, both parsers reject unknown template fields, and dropping the field is not sufficient either because every supported backend rejects a `composed_of` naming a template that does not exist. Jaeger therefore probes for `@custom` and, when it is missing, PUTs an empty body with `?create=true`. A user who creates the template inside that window keeps it: the cluster refuses Jaeger's body rather than replacing theirs, and a refused create is settled by probing again instead of by matching on the error message.

**The priority is fixed at 500.** `indices.spans.priority` tunes the rotation templates on `jaeger-span-*`, which never compete with the exact name `jaeger.spans`, so a value chosen there does not carry over. The test pins the decision rather than the value: a configured `42` must not reach this template.

The RFC's §3.2 prose is updated to match, and Phase 2 milestones 8, 10, 11 and 13 are marked delivered by #8833 and #9363. Milestone 9 stays unticked: this PR renders and creates the templates, but `FactoryBase.createTemplates` does not yet call them, and `RotationConfig.validate` still rejects `data_stream`.

## How was this change tested?

**Unit.** The wire snapshot pins the exact bytes Jaeger sends for all three objects, recorded on every backend version so that collapsing them into one snapshot is itself the assertion that this path never branches on `UsesV8API`. `TestSpanDataStreamComponentsMatchRotationTemplate` is the drift guard: it fails if the two component bodies stop deriving from `jaeger-span.json`. Covered directly: a rendered template missing settings or mappings, unparseable mappings, mappings without `properties`, unmarshalable settings, the fresh-install and existing-`@custom` paths, both outcomes of a refused conditional create, and both transport failure points.

**Live.** `TestElasticsearchStorage_DataStreamTemplates` runs in the ES/OS matrix job and checks the one thing the snapshot cannot: that every supported backend *accepts* these three requests. That matters because the versions in the matrix disagree on exactly the `@custom` question, and because a cluster resolves `composed_of` when the index template is written, so the PUTs succeeding is the compatibility result.

The test stops there. Writing a span through the data stream needs a factory that can be configured for one, which milestone 9 delivers, so the end-to-end write-and-read test lands there rather than here — a test that assembled a writer no production path constructs could not honestly claim to validate the production path.

**Local.** `make fmt`, `make lint` and `make test` are clean, as is `go vet` with the `elasticsearch` tag. `data_stream_template.go` is at 100% statement coverage.

**Deferred, and worth stating explicitly.** Nothing here verifies against a live cluster that `date_nanos` accepts the RFC 3339 string the writer emits. Installing the `@mappings` component proves the cluster parses the mapping, not that it accepts a value: the accepted input formats come from the field's `format` setting, which §3.3 deliberately leaves unset so each engine falls back to its own default. If one of those defaults rejected microsecond-precision RFC 3339, these PUTs would still return 200 and the first real span would fail with a `mapper_parsing_exception`. Only indexing a document tests that pairing. `writer_test.go` pins the format Jaeger emits, so this is the receiving half, and it arrives with the milestone 9 end-to-end test.

## Checklist

- [x] I have read https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully

### CODE REVIEW

**Comment by Me-Priyank** [comment]
@yurishkuro PTAL

File: internal/storage/elasticsearch/esclient/data_stream_template_test.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+	assert.Empty(t, got.DataStream)
+	assert.Equal(t, []string{
+		"prod.jaeger.spans@mappings",
+		"prod.jaeger.spans@settings",
+		"prod.jaeger.spans@custom",
+	}, got.ComposedOf)
+	assert.Equal(t, 500, got.Priority)
+	assert.Equal(t, []string{"prod.jaeger.spans@custom"}, got.IgnoreMissingComponentTemplates)
+}
+
```
Comment (yurishkuro) [inline internal/storage/elasticsearch/esclient/data_stream_template_test.go:172]: Missing wire-snapshot tests. Every other `IndicesClient` operation has a `TestXxxRequestSnapshot` counterpart (e.g. `TestCreateTemplateRequestSnapshot` in `index_client_test.go`) that pins the exact bytes sent to the cluster and stores them in `testdata/`. ADR-012 §Wire-format stability is explicit about this requirement.

The three `PUT` requests issued here — `_component_template/…@mappings`, `_component_template/…@settings`, `_index_template/…` — should each have a snapshot per backend version, following the same pattern as `TestCreateTemplateRequestSnapshot`. Without them, a future refactor of any `render*` function can silently change the wire format with no test catching it.

File: internal/storage/elasticsearch/esclient/data_stream_template.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+	opts := indices.Spans
+	if opts.Replicas == nil {
+		return "", errors.New("span index options have no replica count configured")
+	}
+	out, _ := json.Marshal(map[string]any{
+		"template": map[string]any{
+			"settings": map[string]any{
+				"index.number_of_shards":            opts.Shards,
+				"index.number_of_replicas":          *opts.Replicas,
+				"index.mapping.nested_fields.limit": 50,
```
Comment (yurishkuro) [inline internal/storage/elasticsearch/esclient/data_stream_template.go]: `index.mapping.nested_fields.limit: 50` and `index.requests.cache.enable: true` are duplicated from `jaeger-span.json` (lines 10–11). The PR's "single source of truth" principle covers mappings but not settings: `TestSpanDataStreamMappingsMatchRotationTemplate` guards mappings only. If either constant changes in `jaeger-span.json`, the data-stream `@settings` silently diverges — rotation-path and data-stream-path indices end up with different index settings.

Either derive these from the rendered span template (analogous to how `@mappings` is derived), or add a `TestSpanDataStreamSettingsMatchRotationTemplate` that compares the `settings` stanza the same way the existing drift test compares mappings.

File: internal/storage/elasticsearch/esclient/data_stream_template.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+		},
+	})
+	return string(out), nil
+}
+
+// renderSpanDataStreamIndexTemplate renders the composable index template that
+// declares the span data stream (RFC 0004 §3.2). It composes Jaeger's "@mappings"
+// and "@settings" components with the user-controlled "@custom" one, which is
+// marked ignore_missing so the template is valid before the user creates it.
+func renderSpanDataStreamIndexTemplate(base string) string {
```
Comment (yurishkuro) [inline internal/storage/elasticsearch/esclient/data_stream_template.go:242]: This is the only render function in the file that cannot propagate an error — the `string`-only return forces the marshal error to be discarded. Every other render function (`renderSpanDataStreamMappings`, `renderSpanDataStreamSettings`, `mappingsComponentBody`) returns `(string, error)`.

The comment says "this body is fixed-shape, so marshaling cannot fail", which is true today. But the structural asymmetry means a future developer who adds a field gets silent data truncation: `json.Marshal` returns `nil`, `string(nil)` is `""`, and `putComposableTemplate` issues a PUT with an empty body. Changing the signature to `(string, error)` costs one extra error check in `renderSpanDataStreamTemplates` and removes the gap.

File: internal/storage/elasticsearch/esclient/data_stream_template_test.go (hunk 1 above)
Comment (Me-Priyank) [reply internal/storage/elasticsearch/esclient/data_stream_template_test.go:172]: @yurishkuro  Fixed in 29d1a521. Restored as `TestCreateDataStreamTemplatesRequestSnapshot`, recording all three PUTs across `es.AllVersions`. All six emit identical bytes, so `AssertByVersion` collapses them into one `testdata/create_data_stream_templates.json`. That collapse also asserts this path never branches on `UsesV8API`.

File: internal/storage/elasticsearch/esclient/data_stream_template.go (hunk 2 above)
Comment (Me-Priyank) [reply internal/storage/elasticsearch/esclient/data_stream_template.go]: @yurishkuro  Fixed in 29d1a521. Derived `@settings` from the rendered span template instead of restating it, carried as `json.RawMessage` so it passes through verbatim. Drift test now pins both components.

File: internal/storage/elasticsearch/esclient/data_stream_template.go (hunk 3 above)
Comment (Me-Priyank) [reply internal/storage/elasticsearch/esclient/data_stream_template.go:242]: @yurishkuro  Fixed in 29d1a521. Returns `(string, error)` now. All three bodies marshal through a shared `marshalTemplateBody`, so the error branch is reachable rather than dead.

File: internal/storage/elasticsearch/esclient/data_stream_template.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+// renderSettingsComponent renders the "@settings" component body from the span index
+// template's own settings, carried through as raw JSON. Deriving them rather than
+// restating them is what keeps the shard and replica counts and the fixed index
+// settings from diverging between the rotation and data-stream write paths.
+func renderSettingsComponent(settings json.RawMessage) (string, error) {
+	return marshalTemplateBody("span data stream settings",
+		map[string]any{"template": map[string]any{"settings": settings}})
+}
```
[the remark this answers is not in the document]
Comment (Me-Priyank) [reply internal/storage/elasticsearch/esclient/data_stream_template.go]: `json.Marshal` does validate embedded `RawMessage`: `marshalerEncoder` runs `appendCompact` over the bytes `MarshalJSON` returns, which rejects invalid JSON. `json.Marshal(map[string]any{"template": map[string]any{"settings": json.RawMessage("{not-json")}})` returns:

    json: error calling MarshalJSON for type json.RawMessage: invalid character 'n' looking for beginning of object key string

`TestSpanDataStreamTemplatesErrors/settings_that_cannot_be_marshaled` asserts with `require.ErrorContains`, which fails on a nil error, and it passes. A pre-validating `json.Unmarshal` would duplicate what the encoder already does, so not applying the suggestion.

File: internal/storage/elasticsearch/esclient/testdata/create_data_stream_templates.json [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+  {
+    "method": "PUT",
+    "path": "/_component_template/test.jaeger.spans@settings",
+    "body": {
+      "template": {
+        "settings": {
+          "index.mapping.nested_fields.limit": 50,
+          "index.number_of_replicas": 3,
+          "index.number_of_shards": 3,
+          "index.requests.cache.enable": true
```
[the remark this answers is not in the document]
Comment (Me-Priyank) [reply internal/storage/elasticsearch/esclient/testdata/create_data_stream_templates.json:243]: Intentional, and the PR description was stale. `@settings` reuses the span index template's settings verbatim rather than restating them, which is what removes the duplication flagged earlier in this review. Description updated.

**Comment by Me-Priyank** [comment]
@yurishkuro PTAL

**Comment by Me-Priyank** [comment]
@yurishkuro can you please take a look ?

**Review by yurishkuro — COMMENTED** [review]
Structure looks good, and the earlier points are addressed.

## This slice needs live coverage in the ES/OS matrix

The three PUTs are only exercised against `httptest`. `TestElasticsearchStorage_SyncBulkWriter` and `TestElasticsearchStorage_IndexTemplates` both drive esclient primitives against a real cluster, and `CreateDataStreamTemplates(ctx)` needs no factory wiring to call, so dormancy isn't a reason to defer. It matters because the pinned bytes don't survive most of the matrix: `ignore_missing_component_templates` was added in ES 8.7 and exists in no OpenSearch version, and both parsers reject unknown fields. Dropping the field isn't sufficient either — OpenSearch rejects a `composed_of` naming a template that doesn't exist, so RFC 0004 §3.2's "reference `@custom`, never create it" can't work there; creating an empty `@custom` when absent gives one body valid everywhere, with contents still user-owned.

Have the test write a document, not just assert the templates exist. `@timestamp` is `date_nanos` with no `format`, so numeric input is read as epoch millis, while #8833's write path emits epoch nanos as a decimal string — past the type's 2262 ceiling, document rejected. Keeping nanosecond precision means an RFC 3339 nanosecond string; RFC §3.3's claim that `date_nanos` accepts epoch nanoseconds by default is where the mismatch comes from.

## The neutral-body render is written twice

`renderSpanIndexTemplateBody` re-implements the first half of `RenderIndexTemplate` — replicas check, `ExecuteTemplate`, unmarshal — with its own error wording. Extract that step and call it from both; the data-stream path then takes `inner["settings"]` verbatim and unmarshals `inner["mappings"]`, and `parseSpanIndexTemplateBody` plus two error-path tests go away.

## Smaller points

- `dataStreamTemplatePriority = 500` silently discards `indices.spans.priority`, which `RenderIndexTemplate` honors on ES8. Honor it when non-zero, or say the fixed priority is deliberate.
- The `CreateDataStreamTemplates` comment narrates cross-PR history (#8833, "follow in later slices") — that belongs in the commit message.
- `renderMappingsComponent` writes `@timestamp` into the caller's map; safe today, a trap once anything else parses that body.
- `componentTemplateAPI` / `composableTemplateAPI` differ by two characters for two different endpoints; `indexTemplateAPI` is unambiguous.
- RFC 0004 Phase 2 item 9 is what this delivers, but nothing is marked — here, or once the strategy wires it up?

---

🤖 Drafted with [Claude Code](https://claude.com/claude-code), reviewed by @yurishkuro

**Review by yurishkuro — CHANGES_REQUESTED** [review]
The substance is close now, and CI confirms the matrix fix: `TestElasticsearchStorage_DataStreamTemplates` passes on ES 7.x and OS 1.x, the two versions that rejected the previous body. What I want to flag is the shape of the last two rounds rather than any single item below: each round has fixed exactly the instance named in review, and the fix has arrived carrying a sibling of the same class. Please read the following as four classes to sweep, not six patches to apply.

**Inputs that are silently ignored.** `renderNeutralBody` takes an `innerParams` and then overwrites `params.Shards` / `params.Replicas` from the mapping type's own options (`index_template.go:148-149`), so two of its argument's six fields are inputs that aren't — and the call site passing `innerParams{}` reads as "no shards". This is the class the `priority` comment was about last round, and the priority fix itself has it too: `if priority == 0` treats an explicitly configured `priority: 0` as unset. Take only the fields that vary, and distinguish unset from zero.

**Tests that assert a shape rather than the pairing.** The snapshot test was reworked because it pinned bytes without establishing that a backend accepts them. The integration test added in its place hand-copies the format it is meant to guard — `"@timestamp": start.Format(time.RFC3339Nano)` (`elasticsearch_test.go:255`) — so reverting `writer.go` to epoch nanoseconds leaves it green. Derive both sides from one helper so the test fails when the writer changes.

**Behavior dropped in a refactor.** Extracting `renderNeutralBody` removed the settings-presence check with it, so a neutral body without `settings` now PUTs `{"template":{"settings":null}}` instead of failing locally. Worth checking what else moved out of `RenderIndexTemplate` unnoticed.

**Claims not checked against reality.** The §3.3 correction note gives `date_nanos`'s default as `strict_date_optional_time||epoch_millis`; that is OpenSearch's default, while Elasticsearch's is `strict_date_optional_time_nanos||epoch_millis` (behavior is unaffected — both accept 1–9 fractional digits). And milestone 9 is marked ✅ for `DataStreamStrategy.CreateTemplates()`, which does not exist; this PR delivers the esclient method. A correction note and a milestone tick are load-bearing claims and deserve the same verification as code.

---

🤖 Drafted with [Claude Code](https://claude.com/claude-code), reviewed by @yurishkuro

**Comment by Me-Priyank** [comment]
> The substance is close now, and CI confirms the matrix fix: `TestElasticsearchStorage_DataStreamTemplates` passes on ES 7.x and OS 1.x, the two versions that rejected the previous body. What I want to flag is the shape of the last two rounds rather than any single item below: each round has fixed exactly the instance named in review, and the fix has arrived carrying a sibling of the same class. Please read the following as four classes to sweep, not six patches to apply.
> 
> **Inputs that are silently ignored.** `renderNeutralBody` takes an `innerParams` and then overwrites `params.Shards` / `params.Replicas` from the mapping type's own options (`index_template.go:148-149`), so two of its argument's six fields are inputs that aren't — and the call site passing `innerParams{}` reads as "no shards". This is the class the `priority` comment was about last round, and the priority fix itself has it too: `if priority == 0` treats an explicitly configured `priority: 0` as unset. Take only the fields that vary, and distinguish unset from zero.
> 
> **Tests that assert a shape rather than the pairing.** The snapshot test was reworked because it pinned bytes without establishing that a backend accepts them. The integration test added in its place hand-copies the format it is meant to guard — `"@timestamp": start.Format(time.RFC3339Nano)` (`elasticsearch_test.go:255`) — so reverting `writer.go` to epoch nanoseconds leaves it green. Derive both sides from one helper so the test fails when the writer changes.
> 
> **Behavior dropped in a refactor.** Extracting `renderNeutralBody` removed the settings-presence check with it, so a neutral body without `settings` now PUTs `{"template":{"settings":null}}` instead of failing locally. Worth checking what else moved out of `RenderIndexTemplate` unnoticed.
> 
> **Claims not checked against reality.** The §3.3 correction note gives `date_nanos`'s default as `strict_date_optional_time||epoch_millis`; that is OpenSearch's default, while Elasticsearch's is `strict_date_optional_time_nanos||epoch_millis` (behavior is unaffected — both accept 1–9 fractional digits). And milestone 9 is marked ✅ for `DataStreamStrategy.CreateTemplates()`, which does not exist; this PR delivers the esclient method. A correction note and a milestone tick are load-bearing claims and deserve the same verification as code.
> 
> 🤖 Drafted with [Claude Code](https://claude.com/claude-code), reviewed by @yurishkuro

Thanks @yurishkuro , that framing was the useful part. Swept all four as classes rather than patching the named instances.

**Inputs that are silently ignored.** `renderNeutralBody` now takes a `lifecycleParams` with only the fields that vary between its two callers, and builds `innerParams` itself so shards and replicas come from the mapping type's own options. I checked all four templates first: `IndexPrefix` is consumed only inside a `UseILM` branch, so those four fields really are the lifecycle wiring. `IndexOptions.Priority` is now `*int64`, matching `Replicas` beside it in the same struct, so an explicit `0` is no longer read as unset. `resolvePriority` applies `0` on the ES8 path and `500` for the data stream.

That surfaced one more instance of the same class: `defaultIndexOptions` set `Priority: 0` explicitly, so carrying it over as a pointer to zero would have silently overridden the data stream default on every install. It had to become absent.

**Tests that assert a shape.** A shared helper would not have closed this, since the test never touched `writer.go` and a revert stayed green either way. It now writes through the real `core.SpanWriter` with a `DataStreamRotation`, so the `@timestamp` and op_type the cluster validates are the ones the write path produced. Reverting to epoch nanoseconds gets the item rejected and fails the test.

**Behavior dropped in a refactor.** Restored the settings check, and added the symmetric mappings one. I diffed the old `RenderIndexTemplate` against the new one line by line: the unknown-type check, replica check, render and decode all survived, so the settings check was the only thing lost.

**Claims not checked.** Corrected, and verified against source rather than doc summaries. The note now gives both defaults. Milestone 9 is un-ticked since `DataStreamStrategy.CreateTemplates()` does not exist and this PR delivers the esclient method.

The wire snapshots are unchanged, which is the independent check that the priority work altered no output.

**Review by yurishkuro — CHANGES_REQUESTED** [review]
The substance is in good shape and CI is green across all six backends. My ask this round is structural: this PR bundles three concerns, two of which stand on their own, and I'd like them landed separately before the data-stream work gets another detail round.

The production code here is only ~360 lines — the bulk is a 357-line test and a 259-line generated fixture — so line count isn't the problem. The problem is 17 files spanning three unrelated changes, one of which alters the runtime behavior of already-merged code in a package a reviewer of "add data-stream templates" isn't looking at.

## The split

**PR 1 — `IndexOptions.Priority` becomes `*int64`** (`config.go`, `config_test.go`, `es-rollover/init/flags*.go`, `action_test.go`, `factory_test.go`, `options.go`, `index_client_test.go`). This is a standalone bug fix with nothing to do with data streams: `setDefaultIndexOptions` treats an explicitly configured `priority: 0` as unset and overwrites it with the default. The wire snapshots are the proof that it changes no output.

**PR 2 — `@timestamp` as an RFC 3339 nanosecond string** (`writer.go`, `model.go`, `writer_test.go`, plus the RFC §3.3 correction note). A fix to the write path from #8833, in a different package. Its live proof arrives with PR 3, which is fine — the unit test pins the format meanwhile.

**PR 3 — the data-stream templates**: `data_stream_template.go`, its test, the fixture, the `renderNeutralBody` extraction, the integration test, the RFC §3.2 note and the milestone. Genuinely coupled.

Optionally the `renderNeutralBody` extraction could precede PR 3 on its own, verified by unchanged snapshots, so PR 3 only adds a file. Nice to have, not required.

Below is what each needs, so none of them comes back for a second round on something I could tell you now.

## PR 1 is not ready as it stands

**No regression test for the bug it fixes.** Every `config_test.go` change is mechanical (`10` → `new(int64(10))`). Nothing asserts the new behavior — a target with an explicit `priority: 0` against a source default of 10 must keep 0. Reverting `setDefaultIndexOptions` to `if target.Priority == 0` leaves the suite green.

**The es-rollover flag path destroys the distinction it just created.** `flags.go:35-38` register the priority flags with a default of `0`, so `v.GetInt64(prioritySpanTemplate)` always returns a value and `new(v.GetInt64(...))` always yields a non-nil pointer — pointer-to-zero whenever the user passed no flag. "Unset" becomes "explicitly 0" at the config boundary, which is the same class as the `defaultIndexOptions` instance you just fixed. Today both render priority 0 on the ES8 path, so the answer may simply be to say that in a comment; if the distinction is meant to survive here, note that `v.IsSet` on a pflag-bound key needs verifying before you rely on it.

**`options.go` justifies its change by a feature that isn't in that PR.** "an explicit 0 here would override the data stream's" is the only stated reason for dropping `Priority: 0` from `defaultIndexOptions`, and it disappears if the data-stream priority stays fixed at 500 (see below). Then the removal is a behavior-neutral cleanup that needs a different justification, or no place in PR 1.

`RenderIndexTemplate` also calls `resolvePriority(m.options(indices).Priority, 0)`, recomputing the options `renderNeutralBody` already resolved. And with a fixed 500 there is one caller with a constant fallback, so inline it rather than keeping a two-argument helper.

## PR 2 is essentially ready

I checked the two things that would have made it unsafe. Nothing reads `Span.Timestamp` back — `writer.go:117` is the only reference, and `from_dbmodel.go`/`to_dbmodel.go` touch `Log.Timestamp`, a different field — so the format change is write-only. And `TestWriteSpan_DataStreamTimestamp` does fail on a revert to epoch nanoseconds, so the fix carries a real regression test that does not depend on the templates.

Two one-liners: `spanStartTime.UTC()` is redundant, because `model.EpochMicrosecondsAsTime` already returns UTC (`jaeger-idl/model/v1/time.go:15`), so it guards a case that cannot occur. The test asserts `date.Format(...)` without `.UTC()` and passes only because `date` happens to be UTC, pinning nothing either way. Dropping the `.UTC()` settles both.

## PR 3 carries the review load

**Keep the data-stream priority fixed at 500.** Last round I offered "honor it when non-zero, or say the fixed priority is deliberate," and the first was taken. On reflection the second is right, and `resolvePriority`'s fallback parameter is the tell: one config field now means two different things with two different defaults. The two templates never compete with each other — `jaeger-span-*` versus the exact name `jaeger.spans` — so a user who tuned `spans.priority` down for their rotation templates had a reason that does not transfer, and they now silently get a data-stream template at that value where RFC §3.2 chose 500 to outrank a broad user catch-all. `TestSpanDataStreamIndexTemplatePriority` makes this visible: it asserts that priority 0 reaches the data-stream template, pinning a value that is a misconfiguration. Taking this also decouples PR 1 from the data-stream work entirely.

**The `@custom` TOCTOU from the 7 Aug bot review is still unanswered.** `ensureCustomComponent` GETs then PUTs, so a user writing `@custom` between the two loses it — which contradicts the unqualified "never writes to it again" in the comment. Query parameters are already appended to `endpoint` elsewhere in the package (`index_client.go:70`), so `?create=true` is cheap; the cost is that "already exists" comes back as a 400 carrying an exception type rather than a clean 409, which is uglier to match than the current 404 check. I would accept the race — it is a same-second window on startup — but then say so in the comment instead of stating an absolute. Either way the bot comment needs a reply.

**Two comments assert backend behavior that was not checked.** `componentTemplateExists` says HEAD "is not uniformly available across the supported ES/OpenSearch matrix," while `TestsOnlyTemplateExists` twelve lines away in `index_client.go` probes `_index_template/<name>` with HEAD, and both endpoints arrived together in ES 7.8 with OpenSearch 1.x forking 7.10. Separately, `emptyComponentBody`'s justification says every supported backend requires at least one of settings, mappings or aliases; the integration test proves `{"template":{"settings":{}}}` is accepted, not that `{"template":{}}` is rejected. Verify the claims or drop them. This is the same class as the §3.3 note from last round.

**The `date_nanos` rationale is wrong, though the mapping is right.** `dbmodel.Span.StartTime` is epoch microseconds (`model.go:52`) and `spanStartTime` comes from `model.EpochMicrosecondsAsTime`, so the write path never has nanosecond precision. `renderMappingsComponent` says `@timestamp` is `date_nanos` "to preserve the nanosecond precision Jaeger writes." The real reason is that plain `date` is millisecond precision and would truncate the microseconds, which still justifies `date_nanos`. Fix the reason, keep the type.

**The RFC correction note does not reach every place the RFC states the claim.** The note goes in after §3.2's `@custom` paragraph, but line 159 still says user customizations live in a template "which Jaeger does not touch", and line 175, under "User Customization via `@custom` Pattern", says "Jaeger never creates or modifies this template — it is entirely user-controlled." Both are false in the create-when-absent case, and a reader landing on §3.5 never sees the note.

**Smaller.** In `putComposableTemplate`, `responseError.StatusCode != http.StatusOK` can never be false once `errors.As` succeeds, because `request` only builds a `ResponseError` for a non-200 — it mirrors `CreateTemplate`, so consistency is a defense, but it reads as a condition that means something. And the integration test's `SpanWriter` always emits a service:operation document into `<prefix>jaeger-service-*`, which `TestsOnlyDeleteDataStreamObjects` does not reclaim and which the test never registers `s.CleanUp` for; nothing breaks today because later tests purge at init, but the test claims to own both ends of its cleanup.

## What is solid

I verified the load-bearing claim behind `lifecycleParams{}`: across all four embedded templates, `IndexPrefix` and `IsOpenSearch` appear only inside `{{ if .UseILM }}` branches, so rendering the data-stream body from the zero value cannot drop a setting or leak a rotation alias into a data stream. That claim holds.

The test layering is right now. The snapshot pins what Jaeger sends, the drift test pins that both components derive from `jaeger-span.json`, and the integration test proves a real cluster accepts them and that the writer's own `@timestamp` round-trips — green on ES 7.x and OS 1.x, the pair that rejected the earlier body. Milestone 9 is correctly left un-ticked with a note on what is and is not delivered.

---

🤖 Drafted with [Claude Code](https://claude.com/claude-code), reviewed by @yurishkuro

**Comment by Me-Priyank** [comment]
Got it @yurishkuro, splitting it as described.

Taking PR 1 (`IndexOptions.Priority` as `*int64`) and PR 2 (`@timestamp` as an RFC 3339 nanosecond string) out first as independent PRs, each with the additions you listed: the missing regression test and the es-rollover flag boundary for PR 1, and dropping the redundant `.UTC()` for PR 2.

Once those land I will rebase this branch so it keeps only the data-stream files, and apply the PR 3 round: priority fixed at 500 with the override test removed, the `@custom` TOCTOU and the bot comment, the two unverified backend claims, the `date_nanos` rationale, the two remaining RFC spots, and the service-index cleanup in the integration test.

One more of the same class that you did not name, folding it into that round: the `componentTemplateAPI` comment asserts data streams require ES 7.8+/OS 2.0+, but `TestElasticsearchStorage_DataStreamTemplates` has no version gate and is green on OpenSearch 1.3, so the claim contradicts this PR's own CI.

Marking this draft until the first two are merged.

**Comment by Me-Priyank** [comment]
@yurishkuro PTAL...  rebased and split. #9344 and #9363 removed both sets of hunks, so this is 17 files down to 6 and the conflict is cleared.

All review items applied:

- Priority fixed at 500. The test now pins the decision, not the value: a configured `42` must not reach the template. `resolvePriority` is gone.
- `@custom` comment states the TOCTOU instead of an absolute. Replied on the bot thread.
- Both unverified backend claims dropped rather than restated.
- `date_nanos` rationale corrected, now cites §3.3 instead of restating it.
- Integration test reclaims the service index it writes.
- RFC §3.2 rewritten per #9352, not annotated. Milestone 9 stays unticked.

**Comment by Me-Priyank** [comment]
@yurishkuro could you please take a look ?

**Review by yurishkuro — APPROVED** [review]
